src/util/file-filter.js

```diff
@@ -0,0 +1,109 @@
+/* @flow */
+import path from 'path';
+
+import minimatch from 'minimatch';
+
+import {createLogger} from './logger';
+
+const log = createLogger(__filename);
+
+// Use this function to mimic path.resolve without resolving to absolute path.
+export const normalizeResolve = (file: string): string => {
+  // normalize
+  file = path.normalize(file);
+
+  // trim trailing slash
+  if (path.parse(file).base && file.endsWith(path.sep)) {
+    return file.slice(0, -1);
+  }
+  return file;
+};
+
+// FileFilter types and implementation.
+
+export type FileFilterOptions = {|
+  filesToIgnore?: Array<string>,
+  ignoreFiles?: Array<string>,
+  sourceDir?: string,
+  artifactsDir?: string,
+|};
+
+/*
+ * Allows or ignores files when creating a ZIP archive.
+ */
+export class FileFilter {
+  filesToIgnore: Array<string>;
+  sourceDir: string;
+
+  constructor({
+    filesToIgnore = [
+      '**/*.xpi',
+      '**/*.zip',
+      '**/.*/', // any hidden file and folder
+      '**/node_modules/',
+    ],
+    ignoreFiles = [],
+    sourceDir = '',
+    artifactsDir,
+  }: FileFilterOptions = {}) {
+
+    this.filesToIgnore = [];
+    this.sourceDir = sourceDir;
+
+    this.addToIgnoreList(filesToIgnore);
+    if (ignoreFiles) {
+      this.addToIgnoreList(ignoreFiles);
+    }
+    if (artifactsDir) {
+      this.addToIgnoreList([artifactsDir], true);
+    }
+  }
+
+  /**
+   *  Resolve relative path to absolute path if sourceDir is setted.
+   */
+  resolve(file: string): string {
+    if (this.sourceDir) {
+      return path.resolve(this.sourceDir, file);
+    }
+    return normalizeResolve(file);
+  }
+
+  /**
+   *  Insert more files into filesToIgnore array.
+   */
+  addToIgnoreList(files: Array<string>, isDir: boolean = false) {
+    for (const file of files) {
+      this.filesToIgnore.push(this.resolve(file));
+      // If file is a directory, ignore its content too.
+      if (/([/\\]\.{0,2})$/.test(file) || isDir) {
+        this.filesToIgnore.push(this.resolve(path.join(file, '**/*')));
```

Comment:
I think it makes `FileFilter` less powerful when we insert a new pattern dynamically like this. What if someone doesn't want to match all subdirectories in their custom ignore rules? I don't understand why the original patterns can't just be updated. For example, can't we just call it like this to solve the problem? I.E. change the default patterns to include subdirectory matching:
````js
export class FileFilter {
 constructor({
   filesToIgnore = [
     '**/*.xpi',
     '**/*.zip',
     '**/.*/**/*', // any hidden file and folder
     '**/node_modules/**/*',
   ],
  }) {
    ...
  }
}
````
Or do we need to double them up? I would be fine with this too:

````js
export class FileFilter {
 constructor({
   filesToIgnore = [
     '**/*.xpi',
     '**/*.zip',
     '**/.*', // any hidden file
     '**/.*/**/*', // any hidden folder and its contents
     '**/node_modules/',
     '**/node_modules/**/*', // all sub-directory contents
   ],
  }) {
    ...
  }
}
````